File: algorithms/easy/find_maximum_number_of_string_pairs.ex
defmodule Solution do
  @spec maximum_number_of_string_pairs(words :: [String.t()]) :: integer
  def maximum_number_of_string_pairs(words) do
    mp =
      words
      |> Enum.reduce(%{}, fn word, acc -> Map.update(acc, word, 1, &(&1 + 1)) end)

    words
    |> Enum.reduce({0, mp}, fn word, {ct, mp} ->
      rev = String.reverse(word)
      mp = Map.update!(mp, word, &(&1 - 1))

      {ct + if(Map.get(mp, rev, 0) > 0, do: 1, else: 0), mp}
    end)
    |> elem(0)
  end
end

File: algorithms/easy/find_maximum_number_of_string_pairs.rs
impl Solution {
    pub fn maximum_number_of_string_pairs(words: Vec<String>) -> i32 {
        let mut revs = vec![];
        for w in &words {
            revs.push(w.chars().rev().collect::<String>());
        }
        let mut ct = 0;
        for i in 0..words.len() - 1 {
            for j in i + 1..words.len() {
                if words[i] == revs[j] {
                    ct += 1;
                }
            }
        }
        ct
    }
}
